tf2onnx/function/matrixbandpart.py

```diff
@@ -0,0 +1,82 @@
+
+    # 2: "loop" to generate mask matrix: generate col or row of matrix one by one
+    body_ins = [utils.make_onnx_inputs_outputs("trip", onnx_pb.TensorProto.INT64, []),
+                utils.make_onnx_inputs_outputs("cond", onnx_pb.TensorProto.BOOL, []),
+                utils.make_onnx_inputs_outputs("line", onnx_pb.TensorProto.BOOL, [-1, -1])]
+    body_outs = [utils.make_onnx_inputs_outputs("cond_out", onnx_pb.TensorProto.BOOL, []),
+                 utils.make_onnx_inputs_outputs("line_out", onnx_pb.TensorProto.BOOL, [-1, -1]),
+                 utils.make_onnx_inputs_outputs("res", onnx_pb.TensorProto.BOOL, [-1, -1])]
+    # body graph
+    node_name = utils.make_name("const_zero_bool")
+    const_zero_bool = ctx.make_const(name=node_name, np_val=np.array([[0]]).astype(np.bool))
+    slice_node = ctx.make_node(op_type="Slice", inputs=["line"],
+                               attr={"axes": [counter_axis], "starts": [0], "ends": [-1]})
+    new_line = ctx.make_node(op_type="Concat", inputs=[const_zero_bool.output[0], slice_node.output[0]],
+                             outputs=["line_out"], attr={"axis": counter_axis})
+    body_nodes = list([slice_node.op, new_line.op])
+    body_nodes.append(utils.make_onnx_identity("cond", "cond_out"))
+    body_nodes.append(utils.make_onnx_identity("line", "res"))
+    body_graph = helper.make_graph(body_nodes, utils.make_name("MatrixBandPart_body"), body_ins, body_outs)
+    # initial value of body vars
+    shape = ctx.make_node(op_type="Shape", inputs=[node.input[0]])  # dtype of result is int64
+    node_name = utils.make_name("line_num_index")
+    col_or_row_num_index = ctx.make_const(name=node_name, np_val=np.array(axis).astype(np.int32))
+    line_num = ctx.make_node(op_type="Gather", inputs=[shape.output[0], col_or_row_num_index.output[0]])
+
+    trip_cnt = line_num.output[0]
+    node_name = utils.make_name("true")
+    cond = ctx.make_const(name=node_name, np_val=np.array(1).astype(np.bool)).output[0]
+    col_init = one_line.output[0]
+
+    loop_node = ctx.make_node(op_type="Loop", inputs=[trip_cnt, cond, col_init], output_count=2,
+                              attr={"body": body_graph})
+    # convert generated mask matrix from bool to right shape and data type
+    squeeze = ctx.make_node(op_type="Squeeze", inputs=[loop_node.output[1]], attr={"axes": [squeeze_axis]})
+    cast1 = ctx.make_node(op_type="Cast", inputs=squeeze.output, attr={"to": onnx_pb.TensorProto.FLOAT})
+    mask_matrix = ctx.make_node(op_type="Transpose", inputs=cast1.output) if axis == 1 else squeeze
+    cast2 = ctx.make_node(op_type="Cast", inputs=mask_matrix.output,
+                          attr={"to": ctx.get_dtype(node.input[0])})
+    res = ctx.make_node(op_type="Mul", inputs=[cast2.output[0], node.input[0]],
+                        name=node.name, outputs=node.output)
+    if axis == 1:
+        nodes = [first_col_or_row, first_col_or_row_casted, zero_line, one_line, shape,
```

Comment:
why use the nodes this way, can you add the node into nodes after its creation. this is wired you put all picked up all created node together into the nodes at the end.

Reply:
done.